drv/i2c-devices/src/raa229620a.rs

```diff
@@ -0,0 +1,169 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use core::cell::Cell;
+
+use crate::{
+    pmbus_validate, BadValidation, CurrentSensor, TempSensor, Validate,
+    VoltageSensor,
+};
+use drv_i2c_api::*;
+use pmbus::commands::raa229620a::*;
+use pmbus::commands::CommandCode;
+use pmbus::*;
+use userlib::units::*;
+
+//
+// This is a special rail value that is issued as a PAGE command to enable
+// reading phase current via PHASE + PHASE_CURRENT
+//
+const PHASE_RAIL: u8 = 0x80;
+
+pub struct Raa229620A {
+    device: I2cDevice,
+    rail: u8,
+    mode: Cell<Option<pmbus::VOutModeCommandData>>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    BadRead { cmd: u8, code: ResponseCode },
+    BadWrite { cmd: u8, code: ResponseCode },
+    BadData { cmd: u8 },
+    BadValidation { cmd: u8, code: ResponseCode },
+    InvalidData { err: pmbus::Error },
+}
+
+impl From<BadValidation> for Error {
+    fn from(value: BadValidation) -> Self {
+        Self::BadValidation {
+            cmd: value.cmd,
+            code: value.code,
+        }
+    }
+}
+
+impl From<Error> for ResponseCode {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::BadRead { code, .. } => code,
+            Error::BadWrite { code, .. } => code,
+            Error::BadValidation { code, .. } => code,
+            _ => panic!(),
+        }
+    }
+}
+
+impl From<pmbus::Error> for Error {
+    fn from(err: pmbus::Error) -> Self {
+        Error::InvalidData { err }
+    }
+}
+
+impl Raa229620A {
+    pub fn new(device: &I2cDevice, rail: u8) -> Self {
+        Raa229620A {
+            device: *device,
+            rail,
+            mode: Cell::new(None),
+        }
+    }
+
+    pub fn read_mode(&self) -> Result<pmbus::VOutModeCommandData, Error> {
+        Ok(match self.mode.get() {
+            None => {
+                let mode = pmbus_rail_read!(
+                    self.device,
+                    self.rail,
+                    commands::VOUT_MODE
+                )?;
+                self.mode.set(Some(mode));
+                mode
+            }
+            Some(mode) => mode,
+        })
+    }
+
+    pub fn turn_off(&mut self) -> Result<(), Error> {
+        let mut op = pmbus_rail_read!(self.device, self.rail, OPERATION)?;
+        op.set_on_off_state(OPERATION::OnOffState::Off);
+        pmbus_rail_write!(self.device, self.rail, OPERATION, op)
+    }
+
+    pub fn turn_on(&mut self) -> Result<(), Error> {
+        let mut op = pmbus_rail_read!(self.device, self.rail, OPERATION)?;
+        op.set_on_off_state(OPERATION::OnOffState::On);
+        pmbus_rail_write!(self.device, self.rail, OPERATION, op)
+    }
+
+    pub fn set_vout(&mut self, value: Volts) -> Result<(), Error> {
+        if value > Volts(3.050) {
```

Comment:
This is intended to be `3.050` and not say `3.500`

Reply:
yes, thanks for checking!